lib/atlas/sparse_graph_query.rb

```diff
@@ -0,0 +1,41 @@
+module Atlas
+  class SparseGraphQuery
+    include ActiveDocument
+
+    FILE_SUFFIX = 'gql'
+    DIRECTORY   = 'sparse_graph_queries'
+
+    attribute :graph_key, Symbol
+    attribute :graph_attribute, String
+    attribute :query, String
+
+    validates_presence_of :graph_key, :graph_attribute
+    validates_inclusion_of :graph_attribute, in:
+      GraphValues::VALID_GRAPH_METHODS
+
+    validate :validate_graph_key
+
+    def self.find_by_graph(graph_key, graph_attribute)
```

Comment:
I hadn't considered this, but storing the key and attribute in the document means we don't benefit from the caching provided by `Manager` and involves searching through the whole collection each time. Perhaps it's also duplicating information from the filename?

My first thought was we could put this data in the filename – [like Edge does with the supplier, consumer, and carrier](https://github.com/quintel/atlas/blob/ec45de4edc5e07be658ebeece115252181f86787/lib/atlas/edge.rb#L98-L111) – but that's going to make for some very ugly (and long!) filenames for queries which affect edges.

Reply:
> [..] but that's going to make for some very ugly (and long!) filenames for queries which affect edges.

Yes. That's not to nice. Filenames might become useful - if we could for example `id` all the nodes and edges. So rename all nodes to:

```
<id>.households_final_demand_electricity.final_demand.ad
```

And put that `<id>` in the sparse graph query name aswell:

```
some_query@<id>-demand.ad
```

but this seems a bit radical

---

Another approach might be to whitelist the sparse graph queries to the nodes and edges themselves but that's not to nice, because you get the same tight coupling as with `graph_methods`. 

```
# Content of households_final_demand_electricity.final_demand.ad
...
- sparse_graph_queries.demand = some_query
- sparse_graph_queries.input.electricity = some_query_2
```

This would make it easier to do just a:

```ruby
Atlas::Node.find(graph_key)
  .sparse_graph_queries
  .fetch(graph_attribute) # => some_query
```

And it does help when the node is renamed... 🤔

Reply:
I had another idea. We can also let the query solve which attribute it's going to update. So:

```ruby
- query = UPDATE(<key>, <property>, 5 * DATASET_INPUT(<etlocal_key>))
```

This would mean that *all* the sparse graph queries need to run for which there are DATASET_INPUT's. 💭

Reply:
> Another approach might be to whitelist the sparse graph queries to the nodes and edges themselves but that's not to nice, because you get the same tight coupling as with graph_methods.

Maybe a nice variation on that would be to have two SparseGraphQuery subclasses: one for nodes and one for edges:

```ruby
class SparseGraphQuery::Node < SparseGraphQuery
  attribute :demand
  attribute :number_of_units
  attribute :input
  attribute :output
  # ...
end

class SparseGraphQuery::Edge < SparseGraphQuery
  attribute :parent_share
  attribute :child_share
  # ...
end
```

Now we would have one document per graph object. A node graph query might look like:

```ruby
- demand = some_query
- number_of_units = some_query_2
```

... and the filename simply matches that of the object the queries modify:

* agriculture_chp_engine_biogas.**node**.ad
* buildings_cooling_savings_insulation_cooling-buildings_useful_demand_cooling@cooling.**edge**.ad

Reply:
May I know why you don't like the coupling between the nodes and a sparse graph query (much as the graph methods right now)? 

So this idea:
```
- sparse_graph_queries.demand = some_query
- sparse_graph_queries.input.electricity = some_query_2
```

The reason I'm asking is that I'm also not to much of a fan of the idea sketched above. The reason is that when renaming a node , you'd also need to rename the sparse graph query (and this will probably be forgotten, it's easy to spec though, but still).

Reply:
>The reason I'm asking is that I'm also not to much of a fan of the idea sketched above. The reason is that when renaming a node , you'd also need to rename the sparse graph query (and this will probably be forgotten).

Okay, skip that 😆 this is already validates by checking if the relevant `Node` or `Edge` already exists. What I probably don't like is that all the queries for all the separate attributes are in one file. Also the split between a `Node` sparse graph query or `Edge` sparse graph query doesn't have to be explicitly in the filename as far as I'm concerned as a simple check if the `key` contains an `@`.

Maybe it could also be done as:

- agriculture_chp_engine_biogas.**demand**.ad
- buildings_cooling_savings_insulation_cooling-buildings_useful_demand_cooling@cooling.**parent_share**.ad

This way you can see which attributes the query sets and the file can remain in a format like:

```
- query = SOME_QUERY()
```

Reply:
> Maybe it could also be done as:
>
> * agriculture_chp_engine_biogas.**demand**.ad
> * buildings_cooling_savings_insulation_cooling-buildings_useful_demand_cooling@cooling.**parent_share**.ad

Fair enough. :+1:

The only catch with this *specific* format is that Atlas uses ".something" to declare the the document should be initialized with a subclass. "some_node.final_demand.ad" creates a `Node::FinalDemand`, "some_dataset.derived.ad" creates a `Dataset::Derived`, and so on. It might be better to separate the attribute with something other than a "."

Reply:
> It might be better to separate the attribute with something other than a "."

I found this out the hard way 😅 - so I used a `+` sign.